lib/dom_parser.rb

```diff
@@ -0,0 +1,143 @@
+require_relative 'tree.rb'
+require_relative 'parse_tag.rb'
+
+class DomParser
+  
+  attr_accessor :parse_tree
+
+  def initialize
+    @parse_tree = TagTree.new
+  end
+
+  def dom_reader(html)
+    
+    if File.file?(html)
+       lines = File.readlines(html)
+       html = lines.join.gsub("\n", " ")
+    end
+    build_dom_tree(html)
+
+  end
+    
+  #Start method tokenize to create text and tag tokens
+  def tokenize(html_string)
+    
+    regex = /<.*?>/
+    tag_list = html_string.scan(regex)
+    text_list = html_string.split(regex)
+    
+    #puts "Tag: #{tag_list}"
+    #puts "Text: #{text_list}"
+
+    tag_list_index  = 1
+    text_list_index = 0
+    tokens_array = []
+
+    token = tag_list[0]    #Since we always start with a tag - ASSUME!
+    tokens_array << token
+
+    while tag_list_index < tag_list.size
+      if text_list[text_list_index].strip.empty? 
+        tokens_array << tag_list[tag_list_index]
+      else
+        tokens_array << text_list[text_list_index].strip 
+        tokens_array << tag_list[tag_list_index].strip 
+      end
+      tag_list_index += 1
+      text_list_index += 1
+    end
```

Comment:
I'm not sure, but I think that the index of the text might be thrown off by odd spaces in the text, and that's why sometimes it prints out of order.